infer/environment.h

```diff
@@ -27,6 +27,9 @@ class TypeTestReverseIndex final {
     TypeTestReverseIndex(const TypeTestReverseIndex &rhs) = delete;
     TypeTestReverseIndex(TypeTestReverseIndex &&rhs) = default;
 
+    TypeTestReverseIndex &operator=(const TypeTestReverseIndex &rhs) = delete;
+    TypeTestReverseIndex &operator=(TypeTestReverseIndex &&rhs) = delete;
```

Comment:
Did you want to delete the move assignment as well? This means that you can only ever initialize new values of TypeTestReverseIndex from existing values by initialization with an rvalue.

Reply:
🤷  no opinion. my main motivation is to prevent implicit copies.

Reply:
I think it's fine to leave it deleted until it's needed then 👍